guppylang/std/_internal/compiler/prelude.py

```diff
@@ -151,3 +158,36 @@ def compile_with_inouts(self, args: list[Wire]) -> CallReturnWires:
 
     def compile(self, args: list[Wire]) -> list[Wire]:
         raise InternalGuppyError("Call compile_with_inouts instead")
+
+
+class UnwrapOpCompiler(CustomInoutCallCompiler):
+    """Compiler for operations that require unwrapping an single option value result
+    which can potentially cause an error.
+
+    Args:
+        op: A HUGR operation that outputs an option value.
+        err_msg: The error message to be raised if the unwrapped option is an error.
+    """
+
+    op: Callable[[ht.FunctionType, Inst], ops.DataflowOp]
+    err_msg: str
+
+    def __init__(self, op: Callable[[ht.FunctionType, Inst], ops.DataflowOp], err: str):
+        self.op = op
+        self.err_msg = err
+
+    def compile_with_inouts(self, args: list[Wire]) -> CallReturnWires:
+        assert len(self.ty.output) == 1
+        # To instantiate the op we need a function signature that wraps the output of
+        # the function that is being compiled into a sum type.
+        opt_func_type = ht.FunctionType(
+            input=self.ty.input,
+            output=[ht.Either([error_type()], self.ty.output)],
+        )
+        op = self.op(opt_func_type, self.type_args)
+        option = self.builder.add_op(op, *args)
+        result = build_unwrap(self.builder, option, self.err_msg)
```

Comment:
Also note that this inserts the `unwrap` logic in-place. Maybe we should use the same system we have for `array.__setitem__` etc. to only define this once and emit a call here?

Reply:
Agree that would be better but also should probably be a separate PR since it affects other usages of this as well?

Reply:
Indeed, probably all uses should be refactored. However, to adress the other comment you would need to stop using `build_unwrap` anyways